README.md
# ApocSidebar

[![npm: apoc-sidebar](https://img.shields.io/npm/v/apoc-sidebar.svg)](https://www.npmjs.com/package/apoc-sidebar)
[![CircleCI: nju33/apoc-sidebar](https://circleci.com/gh/nju33/apoc-sidebar.svg?style=svg)](https://circleci.com/gh/nju33/apoc-sidebar)
[![tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
![license: mit](https://img.shields.io/packagist/l/doctrine/orm.svg)

Sidebar without dependenceies

![screenshot](https://github.com/nju33/apoc-sidebar/raw/master/images/screenshot.gif?raw=true)

## Install or Download

```bash
yarn add apoc-sidebar
npm i apoc-sidebar
```

Or access to [releases page](https://github.com/nju33/apoc-sidebar/releases).
Then, download the latest version.

## Demo

[https://nju33.github.io/apoc-sidebar/](https://nju33.github.io/apoc-sidebar/)

## Usage

```html
<!-- Elements that catch events -->
<button id="trigger">...</button>

<!--
  `data-apoc-sidebar-sibling` is required only when type is `lid`.
  It must be a sibling of an element to become a sidebar.
-->
<div data-apoc-sidebar-sibling></div>

<!--
  Specify `width` in CSS.
  Also, can hide the first state by adding `display:none` to the style
-->
<div id="sidebar" style="width:300px;display:none">...</div>

<!-- When reading by itself -->
<script src="/path/tp/apoc-sidebar.js"></script>
```

```js
// es
import ApocSidebar from 'apoc-sidebar';

const sidebar = new ApocSidebar(
  document.getElementById('sidebar'),
  {
    // options

    // default
    type: 'slide',
    // There are other types like this
    // - 'water',
    // - 'push',
    // - 'lid',
    // - 'door',
    // - 'waterfall',
    // - 'waterfallReverse'

    // Parent surrounding `#sidebar` and `[data-apoc-sidebar-sibling]`
    // default
    container: sidebar.parentElement // the parentElement of the `#sidebar`


    // Which side you put on
    // default
    side: 'left', // or 'right'

    // The `transition-timing-function` value of css attached
    // to all relevant elements
    // default (easeInOutQuad)
    // ref: http://easings.net/
    transitionTimingFunction: 'cubic-bezier(0.455, 0.03, 0.515, 0.955)',

    // The `transition-duration` value of css attached
    // to all relevant elements
    // default
    transitionDuration: '.2s',

    // Wall background color
    // default
    wallBackgroundColor: 'rgba(0,0,0,.3)'
  }
);

sidebar.init();
// You can also decide "Open first or close" like `sidebar.init(true)`.


document.getElementById('trigger').addEventListener('click', () => {
  if (sidebar.isOpen()) {
    sidebar.close();
  } else {
    sidebar.open();
  }
});

setTimeout(() => {
  // Delete events
  sidebar.teardown();
}, 999999)
```

### Examples

- `test/fixtures/`
- `example/webpack/`

## LICENSE

The MIT License (MIT)

Copyright (c) 2017 nju33 <nju33.ki@gmail.com>


CHANGELOG.md
## 0.1.0
Farious fixes

## 0.0.1


test/specs/spec.js
import test from 'ava';
import ApocSidebar from '../..';

test('instance', t => {
  const error = t.throws(() => new ApocSidebar());
  t.is(error.message, 'Required element');
});


e2e/__tests__/e2e.js
import {spawn} from 'child_process';
import puppeteer from 'puppeteer';

const wait = () => {
  return new Promise(r => {
    setTimeout(r, 1000);
  });
};

describe('e2e', () => {
  let browser;
  let page;
  let server;

  const startServer = () => {
    let flag = false;

    return new Promise(r => {
      server = spawn('yarn', ['start']);
      server.stdout.on('data', data => {
        if (!flag && data.toString().includes('http://localhost')) {
          flag = true;
          r();
        }
      });
    });
  };

  beforeAll(async () => {
    await startServer();
    browser = await puppeteer.launch();
    page = await browser.newPage();
  });

  afterAll(() => {
    browser.close();
    server.kill();
  });

  test('fixtures', async () => {
    await page.goto('http://localhost:3333');

    async () => {
      await expect(
        age.$eval('#slide', el => el.style.transform)
      ).resolves.stringMatching(/300px, 0px, 0px/);
      await page.click('.apoc-sidebar-wall:first-of-type');
      await wait();

      await expect(
        page.$eval('#slide', el => el.style.transform)
      ).resolves.stringMatching(/0px, 0px, 0px/);
    };
  });
});


test/fixtures/index.js
(() => {
  [
    'slide',
    'water',
    'push',
    'lid',
    'door',
    'waterfall',
    'waterfallReverse'
  ].forEach(type => {
    const s = new ApocSidebar(document.getElementById(type), {type});
    s.init();
    document.getElementById(type + '-trigger').addEventListener('click', () => {
      if (s.isOpen()) {
        s.close();
      } else {
        s.open();
      }
    });
  });
})();


build.js
const fs = require('fs');
const path = require('path');
const gulp = require('gulp');
const less = require('gulp-less');
const postcss = require('gulp-postcss');
const autoprefixer = require('autoprefixer');
const easings = require('postcss-easings');
const modules = require('postcss-modules');
const {rollup} = require('rollup');
const rollupConfig = require('./rollup.config');

const MODULE_NAME = 'ApocSidebar';
const globals = {};

exports.bundleAll = () => {
  return new Promise((resolve, reject) => {
    gulp
      .src('src/styles/style.less')
      .pipe(less())
      .pipe(
        postcss([
          autoprefixer({
            browsers: '> 3%, last 2 versions'
          }),
          easings,
          modules({
            getJSON(cssFileName, json) {
              // const baseName = path.basename(cssFileName) + '.json';
              const baseName = 'class-name.json';
              const jsonFileName = path.join(__dirname, 'lib', baseName);
              fs.writeFileSync(jsonFileName, JSON.stringify(json));
            }
          })
        ])
      )
      .pipe(gulp.dest('lib/'))
      .on('end', () => {
        bundle()
          .then(() => resolve())
          .catch(err => reject(err));
      });
  });
};

exports.bundle = bundle;

function bundle() {
  return new Promise((resolve, reject) => {
    rollup(rollupConfig).then(bundle => {
      rollupConfig.cache = bundle;

      ['iife', 'es', 'umd'].forEach(async format => {
        switch (format) {
          case 'iife': {
            const opts = {
              name: MODULE_NAME,
              globals,
              format,
              banner: rollupConfig.banner
            };
            const destPath = './dist/apoc-sidebar.js';
            try {
              const result = await bundle.generate(opts);
              fs.writeFileSync(destPath, result.code);
            } catch (err) {
              reject(err);
            }
            break;
          }
          case 'umd': {
            const opts = {
              name: 'default',
              globals,
              format
            };
            const destPath = './dist/apoc-sidebar.umd.js';
            try {
              const result = await bundle.generate(opts);
              fs.writeFileSync(destPath, result.code);
            } catch (err) {
              reject(err);
            }
            break;
          }
          case 'es':
          default: {
            const opts = {
              globals,
              format,
              exports: 'default'
            };
            const destPath = './dist/apoc-sidebar.es.js';
            try {
              const result = await bundle.generate(opts);
              fs.writeFileSync(destPath, result.code);
            } catch (err) {
              reject(err);
            }
            break;
          }
        }
      });

      resolve();
    });
  });
}


bs-config.js
const {bundle, bundleAll} = require('./build');

let first = true;

module.exports = {
  ui: false,
  ghostMode: false,
  open: false,
  notify: true,
  files: [
    'e2e/**/*',
    'dist/**/*.js',
    {
      match: ['lib/**/*.+(html|js)'],
      fn() {
        if (first) {
          return;
        }
        bundle().catch(err => console.log(err));
      }
    },
    {
      match: ['src/styles/*.less'],
      fn() {
        if (first) {
          return;
        }
        bundleAll().catch(err => console.log(err));
      }
    }
  ],
  server: {
    baseDir: 'e2e/fixtures',
    routes: {
      '/scripts': `${__dirname}/dist`
    }
  },
  port: 3333,
  browser: 'google chrome',
  injectChanges: true
};

bundleAll()
  .then(() => {
    first = false;
    return;
  })
  .catch(err => console.log(err));


rollup.config.js
const nodeResolver = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');
const babel = require('rollup-plugin-babel');
const svelte = require('rollup-plugin-svelte');
const replace = require('rollup-plugin-replace');
const json = require('rollup-plugin-json');
const string = require('rollup-plugin-string');

const banner = `
/*!
 * Copyright 2017, nju33
 * Released under the MIT License
 * https://github.com/nju33/apoc-sidebar
 */
`.trim();

const nodeEnv = process.env.NODE_ENV || 'development';

module.exports = {
  banner,
  cache: null,
  entry: 'lib/apoc-sidebar.js',
  plugins: [
    nodeResolver({jsnext: true}),
    commonjs({include: 'node_modules/**'}),
    babel({include: 'lib/**/*.js', runtimeHelpers: true}),
    svelte({include: 'lib/components/*.html'}),
    replace({
      'process.env.NODE_ENV': JSON.stringify(nodeEnv)
    }),
    json({include: 'lib/**/*.json'}),
    string({include: 'lib/**/*.css'})
  ]
};


lib/apoc-sidebar.js
const defaultOpts = {
  type: 'slide',
  // type: 'water',
  // type: 'push',
  // type: 'lid',
  // type: 'door',
  // type: 'waterfall',
  // type: 'waterfallReverse',
  side: 'left',
  // side: 'right'
  transitionTimingFunction: 'cubic-bezier(0.455, 0.03, 0.515, 0.955)',
  transitionDuration: '.2s',
  wallBackgroundColor: 'rgba(0,0,0,.3)'
};

export default class ApocSidebar {
  constructor(el = null, opts = defaultOpts) {
    if (el === null) {
      throw new Error('Required element');
    }
    this.wall = null;
    this.el = el;
    this.opts = Object.assign({}, defaultOpts, opts);
    if (typeof this.opts.container === 'undefined') {
      this.opts.container = el.parentElement;
    }

    this.handleTransitionendForWall = this.createTransitionendHandlerForWall.bind(
      this
    );
    this.handleTransitionendForSidebar = this.createTransitionendHandlerForSidebar.bind(
      this
    );
    this.handleTransitionendForOther = this.createTransitionendHandlerForOther.bind(
      this
    );
    this.handleClose = this.close.bind(this);

    this.inited = false;
    this.opened = false;
    this.defaultStyle = {};
  }

  createTransitionendHandlerForWall() {
    Object.assign(this.wall.style, {
      webkitBackfaceVisibility: '',
      backfaceVisibility: '',
      willChange: ''
    });
  }

  createTransitionendHandlerForOther() {
    if (this.opts.type === 'lid' && this.isOpen()) {
      Object.assign(this.el.style, {
        zIndex: 9999
      });
    }
  }

  createTransitionendHandlerForSidebar() {
    if (!this.isOpen()) {
      Object.assign(this.el.style, {
        webkitBackfaceVisibility: '',
        backfaceVisibility: '',
        willChange: '',
        borderRadius: 0,
        zIndex: -9999
      });
    } else {
      Object.assign(this.el.style, {
        webkitBackfaceVisibility: '',
        backfaceVisibility: '',
        willChange: '',
        borderRadius: 0,
        zIndex: 9999
      });
    }
  }

  createWall() {
    this.wall = document.createElement('div');
    Object.assign(this.wall.style, {
      position: 'fixed',
      top: 0,
      left: 0,
      width: '100%',
      height: '100%',
      backgroundColor: this.opts.wallBackgroundColor,
      opacity: 0,
      zIndex: -9998,
      ...this.transitionDecls
    });
    this.wall.className = 'apoc-sidebar-wall';
    document.body.appendChild(this.wall);
    this.wall.addEventListener('click', this.handleClose);
    this.wall.addEventListener(
      'transitionend',
      this.handleTransitionendForWall
    );
  }

  init(isOpen) {
    if (typeof isOpen === 'undefined') {
      isOpen = this.el.style.display !== 'none';
    }

    if (this.inited) {
      return;
    }

    this.inited = true;
    this.createWall();
    this.defaultStyle = Object.assign({}, this.el.style);

    if (this.opts.type === 'push') {
      Object.assign(document.body.style, {
        webkitTransition: '.2s',
        transition: '.2s'
      });
      document.body.addEventListener(
        'transitionend',
        this.handleTransitionendForOther
      );
    } else if (this.opts.type === 'lid') {
      this.siblings.forEach(el => {
        Object.assign(el.style, this.transitionDecls);
      });
    }

    const styles = [this.el.style];
    styles.push({
      display: '',
      height: '100%',
      position: 'fixed',
      zIndex: this.initZIndex,
      top: this.initYPosition,
      [this.opts.side]: this.initXPosition
    });

    if (this.opts.type === 'door') {
      styles.push({
        webkitTransformOrigin: `${this.opts.side} top`,
        transformOrigin: `${this.opts.side} top`,
        ...this.postdoor
      });
    } else if (
      this.opts.type === 'waterfall' ||
      this.opts.type === 'waterfallReverse'
    ) {
      styles.push(this.postwaterfall);
    }

    Object.assign.apply(null, styles);

    if (isOpen) {
      this.open();
    }

    this.el.addEventListener(
      'transitionend',
      this.handleTransitionendForSidebar
    );

    setTimeout(() => {
      Object.assign(this.el.style, this.transitionDecls);
    }, 0);
  }

  get width() {
    return getComputedStyle(this.el).width;
  }

  get initXPosition() {
    if (/^(?:lid|door|waterfall|waterfallReverse)$/.test(this.opts.type)) {
      return 0;
    }
    return `-${this.width}`;
  }

  get initYPosition() {
    if (this.opts.type === 'waterfall') {
      return '-100%';
    } else if (this.opts.type === 'waterfallReverse') {
      return '100%';
    }
    return 0;
  }

  get initZIndex() {
    if (this.opts.type === 'lid') {
      return -9999;
    }
    return '';
  }

  get transitionDecls() {
    return {
      wabkitTransitionTimingFunction: this.opts.transitionTimingFunction,
      transitionTimingFunction: this.opts.transitionTimingFunction,
      webkitTransitionDuration: this.opts.transitionDuration,
      transitionDuration: this.opts.transitionDuration
    };
  }

  get preslide() {
    const size = this.opts.side === 'left' ? this.width : `-${this.width}`;
    return {
      webkitTransform: `translate3d(${size}, 0, 0)`,
      transform: `translate3d(${size}, 0, 0)`,
      zIndex: 9997
    };
  }

  get postslide() {
    return {
      webkitTransform: `translate3d(0, 0, 0)`,
      transform: `translate3d(0, 0, 0)`
    };
  }

  get prewater() {
    if (this.opts.side === 'left') {
      return {
        borderTopRightRadius: '40% 80%',
        borderBottomRightRadius: '40% 80%'
      };
    }
    return {
      borderTopLeftRadius: '40% 80%',
      borderBottomLeftRadius: '40% 80%'
    };
  }

  get postwater() {
    if (this.opts.side === 'left') {
      return {
        borderTopRightRadius: '0 80%',
        borderBottomRightRadius: '0 80%'
      };
    }
    return {
      borderTopLeftRadius: '0 80%',
      borderBottomLeftRadius: '0 80%'
    };
  }

  get predoor() {
    return {
      webkitTransform: 'rotateY(0deg)',
      transform: 'rotateY(0deg)'
    };
  }

  get postdoor() {
    return {
      webkitTransform: 'rotateY(90deg)',
      transform: 'rotateY(90deg)'
    };
  }

  get prewaterfall() {
    return {
      borderBottomLeftRadius: '80% 40%',
      borderBottomRightRadius: '80% 40%',
      webkitTransform: 'translate3d(0, 100%, 0)',
      transform: 'rotateY(0, 100%, 0)'
    };
  }

  get prewaterfallReverse() {
    return {
      borderTopLeftRadius: '80% 40%',
      borderTopRightRadius: '80% 40%',
      webkitTransform: 'translate3d(0, -100%, 0)',
      transform: 'rotateY(0, 100%, 0)'
    };
  }

  get postwaterfall() {
    return {
      webkitTransform: 'translate3d(0, 0, 0)',
      transform: 'rotateY(0, 0, 0)'
    };
  }

  get siblings() {
    const siblings = Array.prototype.slice
      .call(this.opts.container.children)
      .filter(el => el.getAttribute('data-apoc-sidebar-sibling') !== null);
    if (siblings.length === 0) {
      throw new Error(`
In 'push' or 'lid' type,
Required [data-apoc-sidebar-sibling] attr to sibling elements`);
    }
    return siblings;
  }

  isOpen() {
    return this.opened;
  }

  isSlideType() {
    return /^(?:slide|water)$/.test(this.opts.type);
  }

  isWaterType() {
    return this.opts.type === 'water';
  }

  isDoorType() {
    return this.opts.type === 'door';
  }

  isWaterfallType() {
    return this.opts.type === 'waterfall';
  }

  isWaterfallReverseType() {
    return this.opts.type === 'waterfallReverse';
  }

  open() {
    if (this.opened) {
      return;
    }

    this.opened = true;

    Object.assign(this.wall.style, {
      webkitBackfaceVisibility: 'hidden',
      backfaceVisibility: 'hidden',
      willChange: 'opacity',
      zIndex: 9998,
      opacity: 1
    });

    if (this.opts.type === 'push') {
      Object.assign(document.body.style, {
        ...this.preslide
      });
    } else if (this.opts.type === 'lid') {
      this.siblings.forEach(el => {
        Object.assign(el.style, {
          ...this.preslide
        });
      });
    }

    Object.assign(this.el.style, {
      width: this.defaultStyle.width,
      webkitBackfaceVisibility: 'hidden',
      backfaceVisibility: 'hidden',
      willChange: 'transform, border-radius',

      ...(this.isSlideType() ? this.preslide : {}),
      ...(this.isWaterType() ? this.prewater : {}),
      ...(this.isDoorType() ? this.predoor : {}),
      ...(this.isWaterfallType() ? this.prewaterfall : {}),
      ...(this.isWaterfallReverseType() ? this.prewaterfallReverse : {}),

      zIndex: 9999
    });
  }

  close() {
    if (!this.opened) {
      return;
    }

    this.opened = false;

    Object.assign(this.wall.style, {
      webkitBackfaceVisibility: 'hidden',
      backfaceVisibility: 'hidden',
      willChange: 'opacity',
      zIndex: -9998,
      opacity: 0
    });

    if (this.opts.type === 'push') {
      Object.assign(document.body.style, {
        ...this.postslide
      });
    } else if (this.opts.type === 'lid') {
      this.siblings.forEach(el => {
        Object.assign(el.style, {
          ...this.postslide
        });
      });
    }

    Object.assign(this.el.style, {
      webkitBackfaceVisibility: 'hidden',
      backfaceVisibility: 'hidden',
      willChange: 'transform, border-radius',
      zIndex: this.initZIndex,

      ...(this.isSlideType() ? this.postslide : {}),
      ...(this.isWaterType() ? this.postwater : {}),
      ...(this.isDoorType() ? this.postdoor : {}),
      ...(this.isWaterfallType() ? this.postwaterfall : {}),
      ...(this.isWaterfallReverseType() ? this.postwaterfall : {})
    });

    this.opened = false;
  }

  teardown() {
    if (this.isOpen()) {
      this.close();
    }

    if (!this.inited) {
      return;
    }

    this.inited = false;

    if (this.opts.type === 'push') {
      document.body.removeEventListener(
        'transitionend',
        this.handleTransitionendForOther
      );
    } else if (this.opts.type === 'lid') {
      this.siblings.forEach(el => {
        el.removeEventListener(
          'transitionend',
          this.handleTransitionendForOther
        );
      });
    }

    this.el.removeEventListener(
      'transitionend',
      this.handleTransitionendForWall
    );
    this.el.removeEventListener(
      'transitionend',
      this.handleTransitionendForSidebar
    );

    document.body.removeChild(this.wall);
    this.el.style = this.defaultStyle;
  }
}
